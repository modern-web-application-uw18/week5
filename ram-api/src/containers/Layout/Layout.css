.Layout{
    width: 100%;
    display: flex;
    flex-flow: column;
    align-items: center;
}

.MainStage {
    display: flex;
    flex-flow: row wrap;
    justify-content: center;   
}