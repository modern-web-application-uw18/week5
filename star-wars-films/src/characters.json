
{
    "characters": {
      "character": [
        {
            "personID": "Luke Skywalker",
            "thumbnail": "https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcQfx6bpsrceI60IE9et0JVgRC8V6L_BCxKJ892NYBUd69KrnKmR",
            "image": "https://i.pinimg.com/originals/8a/27/47/8a2747689d71083d1c766ffff9f37cc1.jpg"
        },
        {
            "personID": "C-3PO",
            "thumbnail": "http://www.starwarshelmets.com/2007/ANH_HD_3po07.jpg",
            "image": "http://fiona-clarke.com/painting/C3PO~0.jpg"
        },
        {
            "personID": "R2-D2",
            "thumbnail": "https://cdn.thisiswhyimbroke.com/images/life-size-r2d2-figure-640x534.jpg",
            "image": "https://i.pinimg.com/736x/78/c7/43/78c743c8fe20b0e20f0f928ea1470d6e--rd-starwars.jpg"
        },
        {
            "personID": "Darth Vader",
            "thumbnail": "http://digitalspyuk.cdnds.net/15/50/980x490/landscape-1449498579-darth-vader-star-wars.jpg",
            "image": "https://images.fineartamerica.com/images/artworkimages/mediumlarge/1/darth-vader-star-wars-michael-greenaway.jpg"
        },
        {
            "personID": "Leia Organa",
            "thumbnail": "https://vignette.wikia.nocookie.net/starwars/images/8/89/Leia_endorpromo02.jpg/revision/latest/scale-to-width-down/500?cb=20080506155343",
            "image": "http://dtlv.com/files/2015/12/jska_priebe_leia_organa_solo_WEB.jpg" 
        },
        {
            "personID": "Owen Lars",
            "thumbnail": "https://vignette.wikia.nocookie.net/starwars/images/9/91/OwenLarsHS-SWE.jpg/revision/latest?cb=20120428164235",
            "image": "https://1.bp.blogspot.com/-qT_LG7Cycvs/ViQtGwhSX3I/AAAAAAAAERE/cl1CsN5ku1Y/s1600/PhilBrown_UncleOwen.jpg" 
        },
        {
            "personID": "Beru Whitesun lars",
            "thumbnail": "https://vignette.wikia.nocookie.net/headhuntersholosuite/images/0/0c/Beru_Lars_002.jpg/revision/latest?cb=20100625172329",
            "image": "https://i.pinimg.com/originals/2b/bb/71/2bbb712405c574c6ce78730e00464a8e.jpg" 
        },
        {
            "personID": "R5-D4",
            "thumbnail": "https://i.pinimg.com/originals/18/22/af/1822af374fe6fd30facbdf47a3b105e2.jpg",
            "image": "https://pre00.deviantart.net/a196/th/pre/i/2016/076/0/9/r5_d4_by_phraggle-d9vgveq.jpg" 
        },
        {
            "personID": "Biggs Darklighter",
            "thumbnail": "https://vignette.wikia.nocookie.net/starwars/images/0/00/BiggsHS-ANH.png/revision/latest?cb=20130305010406",
            "image": "https://vignette.wikia.nocookie.net/starwars/images/2/20/SWEMP8-FC.jpg/revision/latest?cb=20071108175644" 
        },
        {
            "personID": "Obi-Wan Kenobi",
            "thumbnail": "https://zombiesruineverything.files.wordpress.com/2013/11/obs.jpg",
            "image": "https://ixxidesign.azureedge.net/media/1045227/09-Obi-Wan-Kenobi.jpg?width=498" 
        }
      ]
    }
  }
