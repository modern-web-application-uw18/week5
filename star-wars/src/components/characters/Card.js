import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import './Card.css';

class Card extends Component {

  render() {
    const url = `characters/${this.props.data.url.split('/')[5]}`;

    return (
      <div>
        <Link className='character-list' to={url}>{this.props.data.name}</Link>
      </div>
    );
  }
}

Card.propTypes = {
  data: PropTypes.shape({
      name: PropTypes.string,
      birth_year: PropTypes.string
  })
};

export default Card;
