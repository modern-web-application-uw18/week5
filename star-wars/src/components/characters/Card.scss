.character-list {
  text-align: center;
  display: block;
  margin-bottom: 1em;
}
