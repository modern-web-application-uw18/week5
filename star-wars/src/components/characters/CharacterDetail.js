import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import './CharacterDetail.css';
import Spinner from '../widgets/spinner/Spinner';

class CharacterDetail extends Component {
  constructor(props) {
    super(props);
    this.state = {
      data: [], loading: true, characterId: parseInt(this.props.match.params.characterId, 10)
    };
  }

  nextCharacter = () => {
    this.setState((prevState, props) => {
      return { characterId: prevState.characterId + 1, loading: true };
    }, this.updateApiState);
  }

  previousCharacter = () => {
    this.setState((prevState, props) => {
      return { characterId: prevState.characterId - 1, loading: true };
    }, this.updateApiState);
  }

  updateApiState = () => {
    fetch(`https://swapi.co/api/people/${this.state.characterId}`)

    .then(response => response.json())

    .then(data => {
      this.setState((prevState, props) => {
        return { data: data, loading: false};
      });
    })

    .then(() => console.log(this.state.data))
    .catch(error => console.log(error));
  }

  componentDidMount() {
    this.updateApiState();
  }

  render() {
    return (
      <div>
        {this.state.loading &&
          <Spinner />
        }

        {!this.state.loading &&
          <article className='character-detail'>
            <h1 className='character-detail__header'>
              {this.state.data.name}
            </h1>
            <div className='character-detail__details'>
              <div className='character-detail__spec'>Birth year:</div>
              <div className='character-detail__data'>{this.state.data.birth_year}</div>
            </div>
            <div className='character-detail__details'>
              <div className='character-detail__spec'>Height:</div>
              <div className='character-detail__data'>{this.state.data.height} centimeters</div>
            </div>
            <div className='character-detail__details'>
              <div className='character-detail__spec'>Mass:</div>
              <div className='character-detail__data'>{this.state.data.mass} kilograms</div>
            </div>

            <div className='paging-actions'>
              <Link className='btn' to={`/characters/${this.state.characterId - 1}`} onClick={this.previousCharacter}>Previous</Link>
              <Link className='btn' to={`/characters/${this.state.characterId + 1}`} onClick={this.nextCharacter}>Next</Link>
            </div>
          </article>
        }
      </div>
    );
  }
}

CharacterDetail.propTypes = {
  data: PropTypes.shape({
      name: PropTypes.string,
      birth_year: PropTypes.number,
      height: PropTypes.number,
      mass: PropTypes.number
  })
};

export default CharacterDetail
