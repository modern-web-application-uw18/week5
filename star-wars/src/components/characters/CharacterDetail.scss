.character-detail {
  margin-top: 40px;
}

.character-detail__header {
  font-size: 1.3em;
  text-align: center;
  font-family: 'Righteous', cursive;
  margin-bottom: 50px;
}

.character-detail__details {
  display: flex;
  font-size: 1em;
  justify-content: center;
}

.character-detail__data {
  margin-left: 20px;
}

.character-detail__spec {
  position: relative;
  font-weight: bold;
  margin-bottom: 1em;
}
