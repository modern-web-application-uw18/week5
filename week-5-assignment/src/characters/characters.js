import React, {Component} from 'react';
import { Link } from 'react-router-dom';
import '../App.css'
import 'bootstrap/dist/css/bootstrap.css';

class Characters extends Component {
    constructor(props) {
      super(props);
      this.state = {
        results: {},
        page: 1,
    }
}
    componentWillMount() {
      this.getCharacters();
    }
    
    getCharacters = () => {
      fetch(`https://swapi.co/api/people/?page=${this.state.page}`)
      .then(response => response.json())
      .then(data => {
        this.setState(() => {
          return {
            results: data
          };
        });
      });
    }
    
    next = () => {
      this.setState((prevState) => {
        return { page: prevState.page + 1 };
      }, this.getCharacters);
    }
    
    previous = () => {
      this.setState((prevState) => {
        return { page: prevState.page - 1 };
      }, this.getCharacters);
    }

    
    render() {
      return (
        <div>
            <div className="list-group"> 
          {this.state.results.results && this.state.results.results.map((person, idx) => <li className="list-group-item list-group-item-action" onClick={this.getId} index={idx} key={idx}>
            <Link to={`character/${person.name}`}>{person.name}</Link>
          </li>)}
          </div>
          <button className="btn btn-primary" onClick={this.previous}>Previous</button>
          <button className="btn btn-danger"onClick={this.next}>Next</button>
          </div>
       );
    }
  }

  export default Characters;