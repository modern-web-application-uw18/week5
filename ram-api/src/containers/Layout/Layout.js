import React, { Component } from 'react';
import './Layout.css';
import Header from '../../components/Header/Header';
import MainStageRouter from '../../components/MainStageRouter/MainStageRouter';

class Layout extends Component {

    constructor(props) {
        super(props);


        this.state = {
            searchBoxValue: '',
            favoriteIds: JSON.parse(localStorage.getItem("favoriteIds")) || [],
            characters: []
        }

    }



    sortCharactersAlph = () => {
        let characters = this.state.characters.sort(function (a, b) {
            if (a.name < b.name) return -1;
            if (a.name > b.name) return 1;
            return 0;
        });

        this.setState(prevState => {
            return {
                characters: characters
            }
        })
    }

    componentDidMount() {






        let allCharacters = [];
        let initalJsonPullPromise = new Promise((resolve, reject) => {   //first load the "pages" value from the info object of the character API point
            fetch('https://rickandmortyapi.com/api/character/')
                .then(response => response.json())
                .then(data => resolve({
                    pageCount: data.info.pages,
                    characterCount: data.info.count
                }))                  //resolve the promise with the actual page count and character count
        })
        initalJsonPullPromise.then(resolveObj => {

            let promiseArr = [];
            while (resolveObj.pageCount >= 1) {
                promiseArr.push( //new promise/fetch for each page
                    fetch('https://rickandmortyapi.com/api/character/?page=' + resolveObj.pageCount)
                        .then(response => response.json())
                        .then(data => data.results.forEach(character => allCharacters.push(character)))
                );
                resolveObj.pageCount--;
            }
            Promise.all(promiseArr)
                .then(allCombined => { //pushing individual characters into the allCharacters array above to avoid nested for loops as allCombined is array of arrays...
                    this.setState(prevState => {
                        return {
                            characters: allCharacters
                        }
                    })
                    this.sortCharactersAlph();
                });

        })

    }



    searchHandler = (searchBoxValue) => {
        this.setState((prevState) => {
            return { searchBoxValue: searchBoxValue }
        });
    }

    favAddRemoveHandler = (id) => {


        let prevFavoriteIdsArr = [...this.state.favoriteIds],
            indexOfId = prevFavoriteIdsArr.indexOf(id);

        indexOfId > -1 ? prevFavoriteIdsArr.splice(indexOfId, 1) : prevFavoriteIdsArr.push(id);

        this.setState(prevState => {

            localStorage.setItem('favoriteIds', JSON.stringify(prevFavoriteIdsArr));

            return {
                favoriteIds: prevFavoriteIdsArr
            }
        })



    }


    render() {


        return (
            <div className="Layout">
                <Header searchHandler={this.searchHandler} />
                <MainStageRouter searchBoxValue={this.state.searchBoxValue} favAddRemoveHandler={this.favAddRemoveHandler} characters={this.state.characters} />
            </div>
        )
    }
}

export default Layout;
