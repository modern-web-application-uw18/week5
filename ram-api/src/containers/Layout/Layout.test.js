import React from 'react';
import { mount, shallow} from 'enzyme';
import Layout from './Layout'
import { MemoryRouter } from 'react-router-dom';
import characters from '../../__mocks__/characters.json'

describe('<Layout />', () => {
    it('Should render Layout', () => {

        const wrapper = shallow(
            // <MemoryRouter>
                <Layout />
            // </MemoryRouter>
        );


        expect(wrapper).toMatchSnapshot();
    })
})




