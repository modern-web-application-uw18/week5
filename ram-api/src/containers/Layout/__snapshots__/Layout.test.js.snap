// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Layout /> Should render Layout 1`] = `
<div
  className="Layout"
>
  <withRouter(Header)
    searchHandler={[Function]}
  />
  <MainStageRouter
    characterCount={0}
    characters={Array []}
    favAddRemoveHandler={[Function]}
    pageCount={0}
    searchBoxValue=""
  />
</div>
`;
